pkg/v1/sdk/capabilities/discovery/cluster_gvr.go

```diff
@@ -35,94 +58,239 @@ func (q *QueryGVR) Name() string {
-		for _, v := range q.versions {
-			gvrs = append(gvrs, schema.GroupVersionResource{Group: q.group, Version: v, Resource: q.resource})
+	var unmatched []string
+
+	switch {
+	case q.versions == nil:
+		// q.WithVersions method was omitted and q.versions has not been set.
+		// Any version that matches group and/or resource is considered a match.
+		unmatchedGroupResource, err := q.unmatchedGroupResource(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupResource: %w", err)
+		}
+		if unmatchedGroupResource != "" {
+			unmatched = append(unmatched, unmatchedGroupResource)
+		}
+	case q.resource.String == "":
+		unmatchedGroupVersions, err := q.unmatchedGroupVersions(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupVersions: %w", err)
+		}
+		unmatched = append(unmatched, unmatchedGroupVersions...)
+	case q.resource.String != "":
+		unmatchedGVRs, err := q.unmatchedGroupVersionResources(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupVersionResources: %w", err)
 		}
-	} else {
-		gvrs = append(gvrs, schema.GroupVersionResource{Group: q.group})
+		unmatched = append(unmatched, unmatchedGVRs...)
+	}
+
+	q.unmatchedGVRs = unmatched
+	return len(unmatched) == 0, nil
+}
+
+func (q *QueryGVR) validate(cfg *clusterQueryClientConfig) error {
+	if cfg == nil {
+		return fmt.Errorf("clusterQueryClientConfig must not be nil")
+	}
+
+	if q.resource.IsSet && q.resource.String == "" {
+		return ErrInvalidWithResourceMethodArgument
+	}
+
+	if q.versions != nil && containsEmpty(q.versions) {
+		return ErrInvalidWithVersionsMethodArgument
+	}
+
+	return nil
+}
+
+// containsEmpty looks for an empty string or items containing only white space(s)
+// in the given slice of strings.
+func containsEmpty(strs []string) bool {
+	for _, s := range strs {
+		s = strings.TrimSpace(s)
+		if s == "" {
+			return true
+		}
+	}
+	return false
+}
+
+func (q *QueryGVR) unmatchedGroupResource(cfg *clusterQueryClientConfig) (string, error) {
+	groups, resources, err := cfg.discoveryClientset.ServerGroupsAndResources()
+	if err != nil {
+		return "", fmt.Errorf("discovering server group and resource: %w", err)
+	}
+
+	gvr := schema.GroupVersionResource{
+		Group:    q.group,
+		Resource: q.resource.String,
 	}
 
-	groupList, err := config.discoveryClientset.ServerGroups()
+	if !q.groupsContains(groups, q.group) {
+		return gvr.String(), nil
+	}
+
+	// Resource was not specified, meaning there are no unmatched resources.
+	if q.resource.String == "" {
+		return "", nil
+	}
+
+	rscList := q.resourceListInGroup(resources, q.group)
+	if !q.resourceExists(rscList) {
+		return gvr.String(), nil
+	}
+
+	return "", nil
+}
+
+func (q *QueryGVR) unmatchedGroupVersions(cfg *clusterQueryClientConfig) ([]string, error) {
+	groupList, err := cfg.discoveryClientset.ServerGroups()
 	if err != nil {
-		return false, err
+		return nil, fmt.Errorf("discovering server groups: %w", err)
```

Comment:
Ditto.

Reply:
I would argue that having the word "failed" is redundant. My reasoning is stated in https://github.com/vmware-tanzu/tanzu-framework/pull/1724#discussion_r815194795

Reply:
I'm kind of used to using `failed`, `error occurred` etc when adding the error messages, so, not sure I agree with the thanos style guide(I understand their intentions).. also, to keep it consistent across the project can we continue to use the prefixes in error messages?

Reply:
I did a check for just `Errorf` uses and we seem to be all over the map across the teams and packages within this repo:
https://github.com/vmware-tanzu/tanzu-framework/search?q=errorf

Some that say "error" or "failed", some that say "couldn't do 'x'", some that acknowledge an unexpected state, etc.

While we don't have a cross-repo standard in place we _should_ value consistency in a case like this. I'd recommend finding a few other error messages in the query/discovery/capabilities area of the code and making this look like that.

Reply:
I updated the error message to be prefaced with "failed...".